persistence/src/main/java/edu/unc/lib/dl/persist/services/move/MoveObjectsService.java

```diff
@@ -86,6 +102,35 @@ public String moveObjects(AgentPrincipals agent, PID destinationPid, List<PID> p
         return job.getMoveId();
     }
 
+    private void verifyValidDestination(RepositoryObject destObj, PID pid) {
+        RepositoryObject moveObj = repositoryObjectLoader.getRepositoryObject(pid);
+        if (destObj instanceof AdminUnit) {
+            if (!(moveObj instanceof CollectionObject)) {
+                throw new IllegalArgumentException(
+                        "Object with pid: " + pid + " is not a collection and cannot be added to an admin unit");
+            }
+        } else if (destObj instanceof CollectionObject) {
+            if (!(moveObj instanceof FolderObject) && !(moveObj instanceof WorkObject)) {
+                throw new IllegalArgumentException(
+                        "Object with pid: " + pid + " is not a folder or a work and cannot be added to a collection");
+            }
+        } else if (destObj instanceof FolderObject) {
+            if (!(moveObj instanceof FolderObject) && !(moveObj instanceof WorkObject)) {
+                throw new IllegalArgumentException(
+                        "Object with pid: " + pid + " is not a folder or a work and cannot be added to a folder");
+            }
+        } else if (destObj instanceof WorkObject) {
+            if (!(moveObj instanceof FileObject)) {
+                throw new IllegalArgumentException(
+                        "Object with pid: " + pid + " is not a file and cannot be added to a work");
+            }
+        } else {
+            throw new IllegalArgumentException(
+                    "Object with pid: " + pid + " is not a container");
```

Comment:
how about "Object with pid: \<pid\> cannot be added to destination"

Reply:
yes, much better. the pid is the moving object, not the container :-)